include/swift/AST/ClangModuleLoader.h

```diff
@@ -40,6 +43,66 @@ enum class ClangTypeKind {
   ObjCProtocol,
 };
 
+/// A path for serializing a declaration.
+class StableSerializationPath {
+public:
+  struct ExternalPath {
+    enum StepKind {
+      /// A struct/union/class
+      Struct,
+
+      /// An enum
+      Enum,
+
+      /// A C++ namespace
+      Namespace,
+
+      /// A typedef
+      Typedef,
+
+      /// A typedef's anonymous tag declaration.  Identifier is empty.
+      TypedefAnonDecl,
+
+      /// An Objective-C interface.
+      ObjCInterface,
+
+      /// An Objective-C protocol.
+      ObjCProtocol,
+    };
+
+    static bool requiresIdentifier(StepKind kind) {
+      return kind != StepKind::TypedefAnonDecl;
+    }
+
+    SmallVector<std::pair<StepKind, Identifier>, 2> Path;
+
+    void add(StepKind kind, Identifier name) {
+      Path.push_back({kind, name});
+    }
+  };
+private:
+  TaggedUnion<void, const Decl *, ExternalPath> Union;
+
+public:
+  StableSerializationPath() {}
+  StableSerializationPath(const Decl *d) : Union(d) {}
+  StableSerializationPath(ExternalPath ext) : Union(ext) {}
+
+  explicit operator bool() const { return !Union.empty(); }
+
+  bool isSwiftDecl() const { return Union.isa<const Decl*>(); }
+  const Decl *getSwiftDecl() const {
+    assert(isSwiftDecl());
+    return Union.get<const Decl*>();
+  }
+
+  bool isExternalPath() const { return Union.isa<ExternalPath>(); }
+  const ExternalPath &getExternalPath() const {
+    assert(isExternalPath());
+    return Union.get<ExternalPath>();
+  }
```

Comment:
Thoughts on folding these 4 methods into only 2 which return `Optional`s?

Reply:
I find this pattern of accessors that assert you're working with the right kind of thing generally work better in C++.  Also, `ExternalPath` is a non-trivial type.

Reply:
I don't understand how the fact that `ExternalPath` is a non-trivial type is relevant here... what does that have to do with whether we choose to return an `Optional` or not?

Reply:
Correct.  It would be better to not have an API that can only be implemented by copying non-trivial values out of the container.